notebooks/ch11/gradient_stochastic.ipynb

```diff
@@ -0,0 +1,285 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2018-06-26T05:26:51.230997Z",
+     "start_time": "2018-06-26T05:26:50.094473Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# HIDDEN\n",
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "import seaborn as sns\n",
+    "%matplotlib inline\n",
+    "import ipywidgets as widgets\n",
+    "from ipywidgets import interact, interactive, fixed, interact_manual\n",
+    "import nbinteract as nbi\n",
+    "\n",
+    "sns.set()\n",
+    "sns.set_context('talk')\n",
+    "np.set_printoptions(threshold=20, precision=2, suppress=True)\n",
+    "np.random.seed(42)\n",
+    "pd.options.display.max_rows = 7\n",
+    "pd.options.display.max_columns = 8\n",
+    "pd.set_option('precision', 2)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2018-06-26T05:26:51.264049Z",
+     "start_time": "2018-06-26T05:26:51.234055Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# HIDDEN\n",
+    "tips = sns.load_dataset('tips')\n",
+    "tips['pcttip'] = tips['tip'] / tips['total_bill'] * 100"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2018-06-26T05:26:51.269266Z",
+     "start_time": "2018-06-26T05:26:51.266035Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# HIDDEN\n",
+    "def mse_cost(theta, y_vals):\n",
+    "    return np.mean((y_vals - theta) ** 2)\n",
+    "\n",
+    "def grad_mse_cost(theta, y_vals):\n",
+    "    return -2 * np.mean(y_vals - theta)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## A Brief Review\n",
+    "\n",
+    "We initially modeled the tips dataset by heuristically calculating the $\\hat\\theta$ that minimized the MSE cost function, $L(\\hat\\theta, y) = \\frac{1}{n} \\sum_{i=1}^{n}(y_i - \\hat\\theta)^2$. This involved taking the derivative of the MSE, setting it to zero, and solving for $\\hat\\theta$. We found that the MSE-minimizing $\\hat\\theta$ value was simply the mean of the $y$ values in our dataset.\n",
+    "\n",
+    "But for more complicated models and cost functions, there may not be simple algebraic expressions that yield the cost-minimizing $\\hat\\theta$ values. Our solution was a new method called gradient descent, which iteratively updates $\\hat\\theta$ until it doesn't change between iterations. To complete an iteration of gradient descent, we calculate the following:\n",
+    "\n",
+    "$$\n",
+    "\\hat\\theta_{t+1} = \\hat\\theta_t - \\alpha \\cdot \\nabla_{\\hat\\theta} L(\\hat\\theta, y)\n",
```

Comment:
nit: I think we should specify that the gradient is evaluated at theta = theta_t, since currently the equation states to take the gradient but it should be clear that it is the gradient at theta_t (this is also something I think we should clarify in the other Chapter 11 sections).

Also one thing that could get confusing is the difference between theta and theta_hat, since 11.2 uses theta but we have theta_hat here. Let's keep theta_hat as the optimal theta determined at the end of gradient descent, so I think we should relabel the theta_hats here.

Reply:
@SamLau95 talked about this previously in another comment; I think he intended to switch the thetas in 11.2 to theta_hats? https://github.com/DS-100/textbook/pull/35#discussion_r196255249